# Hola! 😀

Soy **Jaime Paz** y mis amigos me suelen llamar James! – *"Tranquilo, no juego en el Real Madrid"* ⚽😅. Sí, como el futbolista James Rodríguez, pero en versión científico de datos… y mejor no me digas como a él. 😉

Como tu revisor en TripleTen, estoy aquí para ayudarte a pulir tu código y tu forma de trabajar con datos. Si algo necesita un ajuste, no te preocupes: la idea es que cada comentario te acerque más a cómo se trabaja en un entorno profesional y que tu proyecto brille con todo su potencial.

Cada vez que encuentre un detalle importante en tu notebook, te lo señalaré para que puedas corregirlo y seguir creciendo. Si en algún punto no logras resolver algo, también estoy para guiarte en próximos intentos de revisión. 🔁

Es muy importante que, cuando veas mis comentarios en el notebook, **no los muevas, no los modifiques y no los borres**. Así mantenemos un historial claro de lo que ya revisamos y de tus avances. ✅

---

### Formato de Comentarios

Revisaré cuidadosamente tu notebook para asegurar que cumpla con los requisitos y te daré comentarios usando el siguiente formato:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> ✅ - ¡Excelente trabajo! Esta parte está bien implementada y contribuye de forma positiva al análisis o al proyecto. Sigue aplicando estas buenas prácticas en las siguientes secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Esta parte del código funciona, pero se puede mejorar u optimizar. Tal vez sea más claro, más eficiente o más fácil de mantener. Te señalaré ideas para que puedas reforzar esta sección.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error que es necesario corregir para aprobar esta parte. Revisa el comentario con calma, ajústalo y vuelve a intentarlo; es clave para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto. 💪
- **Áreas de mejora:** Sugerencias claras sobre lo que puedes reforzar. 💡
- **Temas adicionales para investigar:** Ideas opcionales que puedes explorar por tu cuenta para seguir creciendo.

Estos temas adicionales no son obligatorios ahora, pero pueden ayudarte a profundizar en el futuro. 📚

---

Si tienes dudas o quieres responder a un comentario específico, puedes usar este formato:

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>

**¡Empecemos!** 🚀

<div class="alert alert-block alert-success">

<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

    

<b>Éxito</b> ✅ - ¡Excelente trabajo! Tu proyecto es excepcional y has logrado cubrir los objetivos del mismo. Abajo he dejado mis comentarios y unas pequeñas recomendaciones para que tomes en cuenta en un futuro y te hagan ¡un máster en Data!

¡Felicidades!    

</div>

# 1. Carga de librerías y datos

In [1]:
# Importamos librerías
# manejo de datos
import pandas as pd
import numpy as np

# preprocesamiento
from sklearn.preprocessing import StandardScaler

# modelos de clasificación
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier

# evaluación y validación
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, classification_report
from numpy.random import default_rng

In [2]:
# Carga de datos
df = pd.read_csv('/datasets/users_behavior.csv')

In [3]:
# Visualización
df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [4]:
# Información y estadísticos
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

Nuestro Dataset está completo.

# 2. Preparación de datos

In [5]:
# separación de features y target
X = df.drop(columns=['is_ultra'])
y = df['is_ultra']

# escalado de variables
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# división train / valid / test (60 / 20 / 20)
X_temp, X_test, y_temp, y_test = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42, stratify=y
)

X_train, X_valid, y_train, y_valid = train_test_split(
    X_temp, y_temp, test_size=0.25, random_state=42, stratify=y_temp
)

# verificación de tamaños
print(
    f'Train: {X_train.shape}, '
    f'Validation: {X_valid.shape}, '
    f'Test: {X_test.shape}'
)

Train: (1928, 4), Validation: (643, 4), Test: (643, 4)


# 3. Modelo 1: Regresión Logística

## 3.1. Ajuste de Hiperparámetros

In [6]:
# Modelo Base
log_reg = LogisticRegression(max_iter=2000, solver='liblinear')

# Grilla de hiperparámetros
log_params = {
    'C':[0.01,0.1,1,10], 
    'class_weight':[None,'balanced']
}

# GridSearch: búsqueda de mejores parámetros
log_grid = GridSearchCV(
    log_reg, 
    log_params, 
    cv=5, 
    scoring='accuracy', 
    n_jobs=-1
)

# Entrenar
log_grid.fit(X_train, y_train)

# Mejor modelo
best_log = log_grid.best_estimator_

# Resultados CV
print('Mejores hiperparámetros:', log_grid.best_params_)
print('Accuracy CV:', log_grid.best_score_)

Mejores hiperparámetros: {'C': 0.01, 'class_weight': None}
Accuracy CV: 0.7510396339411883


## 3.2. Evaluación

In [7]:
# Accuracy en VALIDACIÓN
val_acc_log = accuracy_score(y_valid, best_log.predict(X_valid))
print(f'Accuracy en validación: {val_acc_log:.3f}')

Accuracy en validación: 0.747


In [8]:
# Accuracy + métricas en TEST
test_acc_log = accuracy_score(y_test, best_log.predict(X_test))
print(f'Accuracy en test: {test_acc_log:.3f}')
print(classification_report(y_test, best_log.predict(X_test))) # para ver si el modelo está tomando decisiones correctas

Accuracy en test: 0.740
              precision    recall  f1-score   support

           0       0.73      0.98      0.84       446
           1       0.81      0.20      0.32       197

    accuracy                           0.74       643
   macro avg       0.77      0.59      0.58       643
weighted avg       0.76      0.74      0.68       643



### **Conclusiones parciales - Regresión Logística:**

* La regresión logística fue utilizada como modelo base, alcanzando una accuracy cercana al 75% en validación y test.

* El ajuste de hiperparámetros mediante GridSearch seleccionó un valor bajo de C = 0.01, indicando un modelo fuertemente regularizado.

* El rendimiento es estable entre validación y test, lo que sugiere buena capacidad de generalización.

* Sin embargo, el modelo muestra bajo recall para la clase positiva (1), lo que indica dificultades para detectar correctamente los casos de interés.

* Estos resultados motivan la exploración de modelos más complejos capaces de capturar relaciones no lineales.

# 4. Modelo 2: Bosque Aleatorio

## 4.1. Ajuste de Hiperparámetros

In [9]:
# Modelo base
rf= RandomForestClassifier()

# Grilla de hiperparámetros
rf_params = {
    'n_estimators':[200,500], 
    'max_depth':[None,10,20], 
    'min_samples_leaf':[1,2]
}

# GridSearch
rf_grid = GridSearchCV(
    rf, 
    rf_params, 
    cv=5, 
    scoring='accuracy', 
    n_jobs=-1
)

# Entrenar
rf_grid.fit(X_train, y_train)

# Mejor modelo
best_rf = rf_grid.best_estimator_

# Resultados CV
print('Mejores hiperparámetros:', rf_grid.best_params_)
print('Accuracy CV:', rf_grid.best_score_)

Mejores hiperparámetros: {'max_depth': 10, 'min_samples_leaf': 2, 'n_estimators': 500}
Accuracy CV: 0.8158764551510664


## 4.2. Evaluación

In [10]:
# Accuracy en VALIDACIÓN
val_acc_rf = accuracy_score(y_valid, best_rf.predict(X_valid))
print(f'Accuracy en validación: {val_acc_rf:.3f}')

Accuracy en validación: 0.801


In [11]:
# Accuracy + métricas en TEST
test_acc_rf = accuracy_score(y_test, best_rf.predict(X_test))
print(f'Accuracy en test: {test_acc_rf:.3f}')
print(classification_report(y_test, best_rf.predict(X_test)))

Accuracy en test: 0.820
              precision    recall  f1-score   support

           0       0.82      0.94      0.88       446
           1       0.81      0.54      0.65       197

    accuracy                           0.82       643
   macro avg       0.82      0.74      0.76       643
weighted avg       0.82      0.82      0.81       643



### **Conclusiones parciales - Bosque Aleatorio:**

* El modelo Random Forest mejora significativamente el desempeño respecto a la Regresión Logística, alcanzando accuracy ≈ 0.82 en test.

* El ajuste de hiperparámetros seleccionó un bosque profundo controlado (max_depth = 10, min_samples_leaf = 2), reduciendo el riesgo de overfitting.

* Se observa una mejor capacidad para identificar la clase positiva, con un recall de 0.54 frente al 0.20 de la regresión logística.

* El rendimiento es consistente entre validación y test, indicando buena generalización.

* En conjunto, el Random Forest se posiciona como un candidato sólido a modelo final.

## 5. Modelo 3: Árbol de Decisión

## 5.1. Ajuste de Hiperparámetros 

In [12]:
# Modelo base
dt = DecisionTreeClassifier(random_state=12345)

# Grilla de hiperparámetros
dt_params = {
    'max_depth': [None, 5, 10, 20],
    'min_samples_leaf': [1, 2],
    'min_samples_split': [2, 5, 10]
}

# GridSearch
dt_grid = GridSearchCV(
    dt,
    dt_params,
    cv=5,
    scoring='accuracy',
    n_jobs=-1
)

# Entrenar
dt_grid.fit(X_train, y_train)

# Mejor modelo
best_dt = dt_grid.best_estimator_

# Resultados CV
print('Mejores hiperparámetros:', dt_grid.best_params_)
print('Accuracy CV:', dt_grid.best_score_)

Mejores hiperparámetros: {'max_depth': 5, 'min_samples_leaf': 1, 'min_samples_split': 10}
Accuracy CV: 0.7899347284839512


## 5.2. Evaluación

In [13]:
# Accuracy en VALIDACIÓN
val_acc_dt = accuracy_score(y_valid, best_dt.predict(X_valid))
print(f'Accuracy en validación: {val_acc_dt:.3f}')


Accuracy en validación: 0.793


In [14]:
# Accuracy + métricas en TEST
test_acc_dt = accuracy_score(y_test, best_dt.predict(X_test))
print(f'Accuracy en test: {test_acc_dt:.3f}')

print(classification_report(y_test, best_dt.predict(X_test)))


Accuracy en test: 0.802
              precision    recall  f1-score   support

           0       0.79      0.97      0.87       446
           1       0.85      0.43      0.57       197

    accuracy                           0.80       643
   macro avg       0.82      0.70      0.72       643
weighted avg       0.81      0.80      0.78       643



### **Conclusiones parciales - Árbol de Decisión:**

* El árbol de decisión alcanza un accuracy ≈ 0.80 en test, con desempeño estable entre validación y test.

* La profundidad limitada (max_depth = 5) y un min_samples_split = 10 ayudan a controlar el overfitting.

* El modelo presenta alta precisión y recall para la clase 0, pero dificultades para capturar la clase 1 (recall = 0.43).

* Aunque mejora a la regresión logística, queda por debajo del Random Forest en desempeño global.

* Funciona como un modelo interpretable, pero no como la mejor opción final en términos predictivos.

# 6. Modelo 4: Gradient Boosting

## 6.1. Ajuste de Hiperparámetros

In [15]:
# Modelo base
gb = GradientBoostingClassifier()
gb_params = {
    'n_estimators':[300], 
    'learning_rate':[0.05,0.1], 
    'max_depth':[3,5]
}

# Grilla de hiperparámetros
gb_grid = GridSearchCV(
    gb, 
    gb_params, 
    cv=5, 
    scoring='accuracy', 
    n_jobs=-1
)

# GridSearch
gb_grid.fit(X_train, y_train)

# Entrenar
best_gb = gb_grid.best_estimator_

# Resultados CV
print('Mejores hiperparámetros:', gb_grid.best_params_)
print('Accuracy CV:', gb_grid.best_score_)

Mejores hiperparámetros: {'learning_rate': 0.05, 'max_depth': 3, 'n_estimators': 300}
Accuracy CV: 0.7997927461139896


## 6.2. Evaluación

In [16]:
# Accuracy en VALIDACIÓN
val_acc_gb = accuracy_score(y_valid, best_gb.predict(X_valid))
print(f'Accuracy en validación: {val_acc_gb:.3f}')

Accuracy en validación: 0.787


In [17]:
# Accuracy + métricas en TEST
test_acc_gb = accuracy_score(y_test, best_gb.predict(X_test))

print(f'Accuracy en test: {test_acc_gb:.3f}')
print(classification_report(y_test, best_gb.predict(X_test)))

Accuracy en test: 0.815
              precision    recall  f1-score   support

           0       0.82      0.94      0.88       446
           1       0.80      0.52      0.63       197

    accuracy                           0.81       643
   macro avg       0.81      0.73      0.76       643
weighted avg       0.81      0.81      0.80       643



### **Conclusiones parciales - Gradient Boosting:**

* Gradient Boosting logra un accuracy ≈ 0.82 en test, con resultados consistentes entre CV, validación y test.

* El uso de learning_rate bajo (0.05) y árboles poco profundos (max_depth = 3) favorece una buena generalización.

* Presenta muy buen desempeño en la clase 0 y una mejora clara en recall de la clase 1 respecto a regresión logística y árbol simple.

* Su performance es similar al Random Forest, aunque ligeramente inferior en recall de la clase minoritaria.

* Se posiciona como un modelo competitivo y robusto, aunque no el mejor global del experimento.

# 7. Comparación de Modelos

In [18]:
results = pd.DataFrame({
    'Modelo':['Logistic Regression','Random Forest', 'Decission Tree', 'Gradient Boosting'],
    'Validación':[val_acc_log, val_acc_rf, val_acc_dt, val_acc_gb],
    'Test':[test_acc_log, test_acc_rf, test_acc_dt, test_acc_gb]
})

results

Modelo  Validación      Test
0  Logistic Regression    0.746501  0.740280
1        Random Forest    0.800933  0.819596
2       Decission Tree    0.793157  0.802488
3    Gradient Boosting    0.786936  0.814930

### **Conclusiones Parciales: Comparación de Modelos**

* **Random Forest es el mejor modelo**: obtiene el mayor accuracy en validación y en test, mostrando buena generalización.

* **Gradient Boosting** queda muy cerca, con resultados sólidos y estables, aunque levemente por debajo del Random Forest.

* **Decision Tree mejora** claramente a la Regresión Logística, pero queda limitado frente a los modelos ensemble.

* **Logistic Regression** funciona como buen baseline, pero su desempeño es el más bajo en ambos conjuntos.

* En general, **los modelos ensemble (RF y GB) superan a los modelos más simples**, confirmando que capturan mejor relaciones no lineales.

# 8. Prueba de Cordura del Mejor Modelo

In [19]:
# Seleccionamos el modelo con mejor accuracy en test
best_row = results.loc[results['Test'].idxmax()]
best_model_name = best_row['Modelo']

# Diccionario modelo → objeto entrenado
best_models_dict = {
    'Logistic Regression': best_log,
    'Random Forest': best_rf,
    'Decission Tree': best_dt,
    'Gradient Boosting': best_gb
}

# Recuperamos el mejor modelo
best_model = best_models_dict[best_model_name]

print(f'Mejor modelo según test: {best_model_name}')


Mejor modelo según test: Random Forest


In [20]:
# Test de Cordura
# generador aleatorio reproducible
rng = default_rng(42)

# permutar índices del set de test
perm = rng.permutation(len(X_test))

# predicciones con datos desordenados
predictions_permuted = best_model.predict(X_test[perm])

# accuracy con etiquetas originales
sanity_acc = accuracy_score(y_test, predictions_permuted)

# resultado del sanity check
print(
    f'Mejor modelo: {best_model_name} | '
    f'Accuracy permutado: {sanity_acc:.3f}'
)


Mejor modelo: Random Forest | Accuracy permutado: 0.593


### **Conclusiones parciales: Prueba de Cordura**

* Al permutar el conjunto de test, el accuracy cae de ~0.82 a ~0.59, muy cerca del azar.

* Esto indica que el modelo depende de la relación real entre features y etiquetas, y no de patrones espurios.

* El Random Forest no memoriza el orden de los datos, sino que aprende estructura.

* La prueba de cordura valida que el buen desempeño previo es genuino.

* El modelo supera claramente al azar solo cuando los datos están correctamente alineados, como se espera.

# 9. Conclusión Final

Se evaluaron **cuatro modelos de clasificación** (Logistic Regression, Decision Tree, Random Forest y Gradient Boosting) para predecir la variable objetivo, utilizando una metodología robusta con validación cruzada, set de validación y test independiente.

* El **Random Forest** fue el modelo con mejor desempeño general, alcanzando el mayor accuracy en test (~0.82) y un balance más sólido entre precisión y recall en ambas clases.

* Los modelos basados en **árboles** superaron claramente a la **regresión logística**, lo que sugiere la presencia de **relaciones no lineales** en los datos.

* La **prueba de cordura** confirmó que el desempeño del mejor modelo **no es producto del azar**, validando la **capacidad predictiva real** del enfoque utilizado.

* En función de los resultados, el **Random Forest** se selecciona como **modelo final recomendado**, combinando **buen rendimiento, estabilidad y generalización.**

<div class="alert alert-block alert-success">
<b>Respuesta del revisor</b> <a class="tocSkip"></a><br>

Hola Fiorella, he realizado un checklist técnico para determinar si nos falta algo, pero revisando el notebook contra los criterios, todos los puntos clave están presentes: 
1. cargaste el dataset, ejecutaste
2. el flujo sin errores,
3. hiciste la división 60/20/20 con random_state y stratify,
4. probaste varios modelos con GridSearchCV,
5. evaluaste en test y
6. ejecutaste una prueba de cordura.
  
En resumen: TODO COMPLETADO ✅, Buen trabajo!🚀
</div>


# Conclusiones generales del proyecto

- Nivel general de cumplimiento: en conjunto, incluyes algunos elementos intermedios pero al igual que avanzados (uso consistente de GridSearchCV, ensembles y una prueba de cordura). ✅💡  
- Fortalezas principales:
  - Buen pipeline reproducible: separación train/valid/test con stratify y random_state, y uso de StandardScaler. ✅  
  - Evaluación robusta: grid search para varios modelos, comparación clara y elección justificada del mejor modelo (Random Forest). ✅  
  - Prueba de cordura (permute test) bien implementada, confirma que no es efecto del orden. ✅  
- Áreas de oportunidad y mejora:
  - **RECOMENDADO**Métricas y análisis: incorpora métricas adicionales (ROC‑AUC, F1, curvas de precisión/recall) y visualizaciones (matrices de confusión, curvas ROC) para una evaluación más completa. 📊⚠️  
  - **RECOMENDADO** Interpretabilidad y diagnóstico: añade importancia de features, análisis de errores (qué casos se clasifican mal) y posibles transformaciones/feature engineering. 🔍  
  - **RECOMENDADO** Código y producto final: modulariza funciones (entrenamiento, evaluación), guarda modelos y registra resultados para producción/replicabilidad. 🧩  

- Siguiente paso recomendado: mejora la interpretación del modelo (importancia de variables y análisis de casos mal clasificados) y amplía las métricas usadas para decidir el modelo final; eso te dará argumentos más sólidos para producción. ¡Vas por muy buen camino, sigue así! 🚀